### PR TITLE
Rewrite CheckAOConsistencyWithGpRelationNode() function, for clarity.

### PR DESCRIPTION
**Pull request by hlinnaka**
I couldn't follow its logic. This makes it easier to understand, I hope.

### CODE REVIEW

**Comment by hlinnaka** [comment]
Extracted from the much larger piece of work in PR #2242.

File: src/backend/access/appendonly/aosegfiles.c [hunk 1]
```diff
@@ -2175,117 +2175,89 @@ PrintPgaosegAndGprelationNodeEntries(FileSegInfo **allseginfo, int totalsegs, bo
 }
 
 
+/*
+ * Cross-check the contents of gp_relation_node and the pg_aoseg
+ * table for given relation.
+ */
 void
-CheckAOConsistencyWithGpRelationNode(Snapshot snapshot, Relation rel, int totalsegs, FileSegInfo **allseginfo)
+CheckAOConsistencyWithGpRelationNode(Snapshot snapshot, Relation rel,
```
Comment (ashwinstar) [inline src/backend/access/appendonly/aosegfiles.c]: There exist similar function for CO `CheckCOConsistencyWithGpRelationNode()`, might be worth writing one abstract function for both.

File: src/backend/access/appendonly/aosegfiles.c [hunk 2]
```diff
@@ -2175,117 +2175,89 @@ PrintPgaosegAndGprelationNodeEntries(FileSegInfo **allseginfo, int totalsegs, bo
...
+		aosegMap[fsinfo->segno] = fsinfo;
 	}
 
 	/*
-	 * Check for any extra entries in pg_aoseg which are not present in
-	 * gp_relation_node
+	 * Check that there is a one-to-one correspondence between the two maps.
 	 */
-	while (i < totalsegs)
+	for (i = 0; i <= AOTupleId_MaxSegmentFileNum; i++)
```
Comment (ashwinstar) [inline src/backend/access/appendonly/aosegfiles.c]: Comes with cost of looping for `AOTupleId_MaxSegmentFileNum` even if table has only 1 segfile, can we add fail fast logic please.

File: src/backend/access/appendonly/aosegfiles.c [hunk 3]
```diff
@@ -2175,117 +2175,89 @@ PrintPgaosegAndGprelationNodeEntries(FileSegInfo **allseginfo, int totalsegs, bo
...
+			PrintPgaosegAndGprelationNodeEntries(allseginfo, totalsegs, gprelnodeMap);
 			elog(ERROR, "Missing pg_aoseg entry %d in gp_relation_node for %s",
-				 allseginfo[i]->segno, RelationGetRelationName(rel));
+				 aosegMap[i]->segno, RelationGetRelationName(rel));
+		}
+
+		/*
+		 * Special case for segment 0: It's ok if the pg_aoseg entry doesn't exist.
+		 * When a new table is created, we create a gp_relation_node entry for
+		 * segment 0, and the underlying empry file, but we don't insert a row in
```
Comment (danielgustafsson) [inline src/backend/access/appendonly/aosegfiles.c]: s/empry/empty/

**Comment by danielgustafsson** [comment]
Any updates on this addressing the review comments?

**Comment by hlinnaka** [comment]
This code was removed, along with perstistent tables, by the big walreplication merge. Hence this is obsolete, closing.

**Pull request closed without merging**